File: Chapter16/SampleViewAnimation/Java/SampleViewAnimation/app/src/main/java/com/ai/android/ExerciseSystemIntents/ViewAnimation1.java
package com.ai.android.ExerciseSystemIntents;

import android.graphics.Camera;
import android.graphics.Matrix;
import android.util.Log;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.view.animation.Transformation;

public class ViewAnimation1 extends Animation {

    private static final String TAG = ViewAnimation1.class.getSimpleName();

    private float centerX;
    private float centerY;
    private Camera cam;

    public ViewAnimation1(float cx, float cy) {
        cam = new Camera();
        centerX = cx;
        centerY = cy;
    }

    @Override
    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        super.initialize(width, height, parentWidth, parentHeight);
        Log.d(TAG, "width: " + width);
        Log.d(TAG, "height: " + height);
        Log.d(TAG, "pwidth: " + parentWidth);
        Log.d(TAG, "pheight: " + parentHeight);
        setDuration(2500);
        setFillAfter(true);
        setInterpolator(new LinearInterpolator());
    }

    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t) {
        applyTransformationNew(interpolatedTime, t);
    }

    protected void applyTransformationNew(float interpolatedTime, Transformation t) {
        Log.d(TAG, "transform: " + interpolatedTime);
        final Matrix matrix = t.getMatrix();
        cam.save();
        cam.translate(0.0f, 0.0f, (1300 - 1300.0f * interpolatedTime));
        cam.rotateY(360 * interpolatedTime);
        cam.getMatrix(matrix);

        matrix.preTranslate(-centerX, -centerY);
        matrix.postTranslate(centerX, centerY);
        cam.restore();
    }
}

File: Chapter16/SampleViewAnimation/Kotlin/SampleViewAnimation/app/src/main/java/com/ai/android/sampleviewanimation/ViewAnimation1.kt
package com.ai.android.sampleviewanimation

import android.graphics.Camera
import android.util.Log
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.Transformation

class ViewAnimation1(
    private val centerX: Float,
    private val centerY: Float
): Animation() {

    private val cam = Camera()

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        super.initialize(width, height, parentWidth, parentHeight)
        Log.d(TAG, "width: $width")
        Log.d(TAG, "height: $height")
        Log.d(TAG, "pwidth: $parentWidth")
        Log.d(TAG, "pheight: $parentHeight")
        duration = 2500
        fillAfter = true
        interpolator = LinearInterpolator()
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation?) {
        applyTransformationNew(interpolatedTime, t)
    }

    private fun applyTransformationNew(interpolatedTime: Float, t: Transformation?) {
        Log.d(TAG, "transform: $interpolatedTime")
        t?.apply {
            matrix?.apply {
                cam.save()
                cam.translate(0.0f, 0.0f, (1300 - 1300.0f * interpolatedTime))
                cam.rotateY(360 * interpolatedTime)
                cam.getMatrix(this)

                preTranslate(-centerX, -centerY)
                postTranslate(centerX, centerY)
                cam.restore()
            }
        }
    }

    companion object {
        const val TAG = "ViewAnimation1"
    }
}